src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionListener.java

```diff
@@ -0,0 +1,30 @@
+package org.jenkinsci.plugins.workflow.flow;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+
+/**
+ * Listens for significant status updates for a {@link FlowExecution}, such as started running or completed.
+ *
+ * @since 2.14
+ * @author Andrew Bayer
+ */
+public abstract class FlowExecutionListener implements ExtensionPoint {
+
+    /**
+     * Called when a {@link FlowExecution} has started running or resumed.
+     *
+     * @param execution The {@link FlowExecution} that has started running or resumed.
+     */
+    public void onRunning(FlowExecution execution) {
+    }
+
+    /**
+     * Called when a {@link FlowExecution} has completed.
+     *
+     * @param execution The {@link FlowExecution} that has completed.
```

Comment:
> it's getting called in a separate thread

`GraphL` is not `Synchronous`, so it will be called some time after the `FlowEndNode` is actually attached. `Synchronous` listeners would be notified of it before `finish`; asynch ones either before or after, arbitrarily, according to when the listener was attached.

This is why I suggested in the other PR that the firing would be better done from `workflow-job`, from `workflow-cps`. It is then more straightforward to reason about exactly when the events are fired.